src/consensus.rs

```diff
@@ -153,41 +167,48 @@ impl<T: Proposition> Consensus<T> {
                 "[MBR-{}] Detected super majority over super majorities: {proposals:?}",
                 self.id()
             );
+            let votes = crate::vote::simplify_votes(&self.votes.values().cloned().collect());
             let decision = Decision {
-                votes: self.votes.values().cloned().collect(),
+                votes,
                 proposals,
                 faults: self.faults(),
             };
-            self.decision = Some(decision.clone());
-            return Ok(VoteResponse::Decided(decision));
+            self.decision = Some(decision);
+            return Ok(VoteResponse::WaitingForMoreVotes);
         }
 
         if self.is_split_vote(&self.votes.values().cloned().collect()) {
             info!("[MBR-{}] Detected split vote", self.id());
             let merge_vote = Vote {
-                gen,
+                gen: signed_vote.vote.gen,
                 ballot: Ballot::Merge(self.votes.values().cloned().collect()).simplify(),
                 faults: self.faults(),
             };
             let signed_merge_vote = self.sign_vote(merge_vote)?;
+            let current_count = self.count_votes(&self.votes.values().cloned().collect());
+            let merge_count = self.count_votes(
+                &signed_merge_vote
+                    .unpack_votes()
+                    .into_iter()
+                    .cloned()
+                    .collect(),
+            );
 
-            if let Some(our_vote) = self.votes.get(&self.id()) {
-                let proposals_we_voted_for = our_vote.proposals();
-                let proposals_we_would_vote_for = signed_merge_vote.proposals();
-
-                if proposals_we_voted_for == proposals_we_would_vote_for
-                    && our_vote.vote.faults.len() == signed_merge_vote.vote.faults.len()
-                {
-                    info!(
-                        "[MBR-{}] This vote didn't add new information, waiting for more votes...",
-                        self.id()
-                    );
-                    return Ok(VoteResponse::WaitingForMoreVotes);
+            let resp = if current_count != merge_count {
+                info!("[MBR-{}] broadcasting merge.", self.id());
+                match self.handle_signed_vote(signed_merge_vote.clone())? {
```

Comment:
We now recursively handle our own vote, this is necessary now that we are doing proper idempotent checks. (previously we would log that we voted and wait for the broadcast to reach us with our own vote)